### ISSUE TITLE
Grammar for the Ltac chapter

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#9447]
I used the grammar extraction/editing tool I'm creating for #9310 to extract the Ltac grammar.  This is a usability test case for the tool.

The Ltac grammar is 2 1/2 pages (see below).  That seems like an intimidating amount of text to put at the beginning of the Ltac chapter.

I could separate the `tactic_expr[0-5]` so it's less overwhelming.  But I don't see an obvious logic in how various features such as `do` and `match` are associated with the `tactic_expr[0-5]`.  Nor is it clear to me that similar features are at the same level.  (Ideally, features should be described in the order in which they appear in the grammar.  That may be problematic.)  Maybe a summary-of-features table should appear near the beginning of the section?  The matching syntax is complex, will need extra work to make it clear.  And it seems like the grammar will need considerable editing to make it easy to read.

The manual's grammar is not so similar to mine.  A bunch of productions are missing or wrong in the manual's grammar.  And a few things are documented that don't seem to exist in my grammar.

@Zimmi48, what are your thoughts?  How do we best use the grammar for Ltac?

I'm not planning to rewrite the Ltac chapter just yet.

**Comment by Zimmi48** [comment]
Some immediate simplification that come to mind:

- inlining some stuff, for instance `Constr.ident`, `lconstr_pattern`, `failkw`, `input_fun`, `tactic_then_locality`, `tactic_then_last`, `selector`, `range_selector_or_nth`, `constr_eval`, `fresh_id`, `tactic_expr4`, `tactic_expr0`, `tactic_atom` (unless these last three are important for some other reason);

- splitting out the `match` grammar.

Overall, it seems to me that there is not much point in having the user manual lie to the users on the complexity of the Ltac grammar. If we had been more thorough in the documentation so far, maybe this would have prevented the grammar from becoming so big. Ltac2, which we are going to promote more and more (when is it going to be merged into Coq @ppedrot @maximedenes?), is going to have a much lighter grammar, I think.

Finally, something needs to be done regarding these `OPT`, `LIST1`, `LIST0`, `SEP`... Do we continue emulating them using square brackets and `...` or do we try to go towards the model of the SSR chapter which is using `prodn` systematically? Or can we have a combination of both where we continue to have lighter grammars *à la* `productionlist` but with support for some Coq Sphinx syntax? @cpitclaudel I'd like your opinion on this.

**Comment by JasonGross** [comment]
Note that the various tactic_expr levels are accessible from tactic notations, as, e.g., tactic3(tac).  So it might be good to keep them separate.

Separately, I think `tryif` has the wrong level; see https://github.com/coq/coq/issues/5441 .  It should be tactic_expr3.

**Comment by cpitclaudel** [comment]
> @cpitclaudel I'd like your opinion on this.

I don't have a strong opinion. I'll admit to not usually paying enormous attention to grammars, but I find it convenient from time to time to be able to click a particular identifier and get documentation.  I think the OCaml manual doesn't look too bad on trhat front (e.g. https://caml.inria.fr/pub/docs/manual-ocaml/modtypes.html).  I think emulating lists with … is fine.

**Comment by jfehrle** [comment]
>  Some immediate simplification that come to mind

Sure.  Haven't done most of the simplification yet.  But even so, I think
the result will be more than 2 pages, not too easy to read and not a great
way to begin the Ltac chapter.  Maybe most of the grammar should go at the
end of the chapter.

> there is not much point in having the user manual lie to the users on the
complexity of the Ltac grammar.

Not something I would propose.

> Finally, something needs to be done regarding these OPT, LIST1, LIST0, SEP...
Do we continue emulating them using square brackets and ... or do we try to
go towards the model of the SSR chapter which is using prodn systematically?
Or can we have a combination of both where we continue to have lighter
grammars *à la* productionlist but with support for some Coq Sphinx syntax?

prodn is useful and readable when the syntax being shown is fairly close to
being a regular expression.  It makes sense for introducing/describing each
Ltac feature, or for each tactic and command.  But the document should also
include the entire grammar as productionlists, probably split into chunks
in the chapters plus an appendix with the full grammar in one place.  (The
grammar for simple_tactic is over 400 lines, BTW.)

Having metacharacters in productionlists such as "[ abc | xyz ]" for an
optional "abc" or "xyz" is confusing because that uses [ | and ] both as
metacharacters and literal strings.  Using "..." to show repetition doesn't
distinguish well between LIST0 and LIST1 without additional [ ] around the
LIST0 item.  Also so similar to ".." that our grammar uses.

I think the best way to handle these is to eschew metacharacters and make
these new non-terminals in the documentation:
LIST0 foo SEP "," becomes

foo_list_opt : foo_list_opt "," foo
             : empty

LIST1 foo SEP "," becomes

foo_list : foo_list "," foo
         : foo

OPT foo becomes

foo_opt : foo
        : empty

This will make the grammar somewhat larger, but I think the greater clarity
makes it a net gain.

**Comment by jfehrle** [comment]
(reposted because email replies don't support markdown :-( )

> Some immediate simplification that come to mind

Sure.  Haven't done most of the simplification yet.  But even so, I think
the result will be more than 2 pages, not too easy to read and not a great
way to begin the Ltac chapter.  Maybe most of the grammar should go at the
end of the chapter.

> there is not much point in having the user manual lie to the users on the complexity of the Ltac grammar.

Not something I would propose.

> Finally, something needs to be done regarding these OPT, LIST1, LIST0, SEP... Do we continue emulating them using square brackets and ... or do we try to go towards the model of the SSR chapter which is using prodn systematically? Or can we have a combination of both where we continue to have lighter grammars *à la* productionlist but with support for some Coq Sphinx syntax?

prodn is useful and readable when the syntax being shown is fairly close to
being a regular expression.  It makes sense for introducing/describing each
Ltac feature, or for each tactic and command.  But the document should also
include the entire grammar as productionlists, probably split into chunks
in the chapters plus an appendix with the full grammar in one place.  (The
grammar for simple_tactic is over 400 lines, BTW.)

Having metacharacters in productionlists such as "[ abc | xyz ]" for 
"abc" or "xyz" is confusing because that uses [ | and ] both as
metacharacters and literal strings.  Using "..." to show repetition doesn't
distinguish well between LIST0 and LIST1 without additional [ ] around the
LIST0 item.  Also so similar to ".." that our grammar uses.

I think the best way to handle these is to eschew metacharacters and make
these new non-terminals in the documentation:
`LIST0 foo SEP ","` becomes

```
foo_list_opt : foo_list_opt "," foo
             : empty
```

`LIST1 foo SEP ","` becomes

```
foo_list : foo_list "," foo
         : foo
```

`OPT foo` becomes

```
foo_opt : foo
        : empty
```

This will make the grammar somewhat larger, but I think the greater clarity
makes it a net gain.

**Comment by Zimmi48** [comment]
Another solution would be to have the meta characters show in a different font than the rest of the text. But I'm also fine with your solution and putting the grammars at the end of chapters for reference does indeed make perfect sense.

**Comment by jfehrle** [comment]
> Another solution would be to have the meta characters show in a different font than the rest of the text.

I had looked at that, however Sphinx doesn't support it.  And I think I prefer the approach I described even if Sphinx did support that.

> The definition can contain token names which are marked as interpreted text (e.g. sum ::= `integer` "+" `integer`) – this generates cross-references to the productions of these tokens. Outside of the production list, you can reference to token productions using token.
>
> **Note that no further reST parsing is done in the production, so that you don’t have to escape * or | characters.**


https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#grammar-production-displays

**Comment by jfehrle** [comment]
My latest thoughts:  It's not too hard to generate an accurate grammar from the .mlg files.  Automatically generating a grammar that has its nonterminals and productions in an order that's great for human consumption is difficult to impossible.  We should reorder the nonterminals and productions in the .mlg files to make them as close as possible to the order we want in the documentation grammar, if only for our own sanity when we want to compare the doc grammar to the .mlgs.  However, the .mlgs are in 30+ files, so they don't provide an overall order.  Also a very few productions can't be reordered in the .mlgs because that will alter how some inputs are parsed.

The best I've come up with is something like this:
1.  `make` automatically generates a file "fullGrammar" with the grammar in the .mlg file order (loading the .mlgs in a defined order, e.g. alphabetical by filename).
2.  A developer copies this file to "orderedGrammar" and reorders it as desired.
3.  A utility invoked from the command line reads orderedGrammar, verifies that only the ordering differs compared to fullGrammar, applies some edits (e.g. expanding LIST1 to a new non-terminal), and generates productionlists and prodns for various subsets of the productions.  This generates an output file "docGrammar" that contains the final documentation grammar and productionlist/prodn for each of the subsets.
4. The Zimmermann utility (to be written) automatically inserts each subset in the right spot in the .rst files.

Each of these files would be checked in to git.  fullGrammar should make it easy to find .mlg modifications.  docGrammar is probably preserved only for its reference value--not to be edited directly.

fullGrammar and orderedGrammar could have exactly the format of .mlg files so I can reuse the .mlg parser from coqpp.  Like this:

```
GRAMMAR EXTEND Gram
  (* added productions *)

  empty: [ [ -> { } ] ] ;

  lpar_id_coloneq:
    [ [ "(" IDENT; ":=" -> { }
    ] ] ;
```
On the other hand, this has considerable clutter that's not needed for the syntax alone.  Perhaps it would not be hard to tweak the coqpp parser to accept something more readable like:

```
MAGIC_TOKEN
  (* added productions *)

  empty: 

  lpar_id_coloneq: "(" IDENT ":="
```

The editing operations I mentioned include global changes such as:
- removing symbols that are only for lookahead, e.g. `test_bracket_ident`
- renaming a few symbols, e.g. `Constr.constr` -> `constr`
- splicing productions for a specified nonterminals into the references to the nt, then removing the nt
(good for prodn)
- Replacing LIST* and OPT constructs with new non-terminals (good for productionlist)

There are many productions in the .mlgs that are not ideal for human consumption.  As much as possible, these should be fixed by modifying the .mlg files.  There appear to be quite a few such edits needed.  For example, these productions should be combined into one production with `OPT "reverse"`:

```
tactic:   ...
      [ match_key "goal" "with" match_context_list "end" -> { }
      | match_key "reverse" "goal" "with" match_context_list "end" -> { }
```

For things that can't be changed in the .mlg files, I have a file that has instructions to edit specific productions if necessary.  (These are tedious, maybe not easy to follow.  I thought this might be better than editing docGrammar directly.):

```
  simple_tactic:
  [ [ DELETE "intros" -> { }
    | REPLACE "intros" ne_intropatterns -> { }
    | WITH "intros" intropattern_list_opt -> { }
    | DELETE "eintros" -> { }
    | REPLACE "eintros" ne_intropatterns -> { }
    | WITH "eintros" intropattern_list_opt -> { }
  ] ] ;
```

A few other thoughts:
- It should be possible to automatically mark productions that come from plugins with an annotation visible in the doc such as `(SSR plugin)`
- We should consider consolidating some of the .mlg files.  Some of them have very little in them.
- Most likely, cmd's and tacn's will show the full syntax; cmdv's tacv's will not be generated automatically.

The reordering and editing of the .mlgs involves a lot of trivial refactoring.  I hope others won't object to this since it will be so helpful for the documentation.

@Zimmi48 let me know your thoughts.

**Comment by Zimmi48** [comment]
This is a grand plan that I like very much. In particular, I think refactoring the parsing files to be closer to what is documented (and the converse) is a very good thing because it will make contributing easier. It would also make sense to move things around so that we strive to have a single grammar (as documented) in a single `mlg` file. I would also be open to changing the coqpp parser to move away from camlp5 syntax and to be more readable (but it has to be a decision of @coq/parsing-maintainers, especially given that it could also affect external plugins).

What would be concrete steps towards concertizing this plan? Maybe we can create a GitHub project to track the various associated issues / PRs.

>The Zimmermann utility

Huh?

**Comment by jfehrle** [comment]
We can't go to a single .mlg file.  Each optionally-loaded plugin may need its own .mlg file.  The .mlg files are almost 11000 lines, much too big for a single file.  But we can organize the grammar into a smaller number of chunks, e.g. one for `constr`, one for vernac, one for tactics, etc.  I think this has already been done to some extent.  But we should be able to merge the smaller mlgs into the others.  Here is the number of nonterminals and productions defined in each mlg.  Items above the line are always loaded (and shown in the order they are loaded).  Items below the line are for additional plugins:

```
parsing/g_constr.mlg:  74 nts,  156 prods
parsing/g_prim.mlg:  23 nts,  30 prods
vernac/g_vernac.mlg:  111 nts,  420 prods
vernac/g_proofs.mlg:  6 nts,  56 prods
toplevel/g_toplevel.mlg:  1 nts,  4 prods
plugins/ltac/extraargs.mlg:  12 nts,  22 prods
plugins/ltac/g_obligations.mlg:  11 nts,  24 prods
plugins/ltac/coretactics.mlg:  45 nts,  65 prods
plugins/ltac/extratactics.mlg:  79 nts,  115 prods
plugins/ltac/profile_ltac_tactics.mlg:  9 nts,  13 prods
plugins/ltac/g_auto.mlg:  24 nts,  37 prods
plugins/ltac/g_class.mlg:  11 nts,  16 prods
plugins/ltac/g_rewrite.mlg:  17 nts,  62 prods
plugins/ltac/g_eqdecide.mlg:  2 nts,  2 prods
plugins/ltac/g_tactic.mlg:  57 nts,  197 prods
plugins/ltac/g_ltac.mlg:  57 nts,  133 prods
plugins/syntax/g_string.mlg:  1 nts,  1 prods
plugins/btauto/g_btauto.mlg:  1 nts,  1 prods
plugins/rtauto/g_rtauto.mlg:  1 nts,  1 prods
plugins/cc/g_congruence.mlg:  2 nts,  5 prods
plugins/firstorder/g_ground.mlg:  5 nts,  10 prods
plugins/syntax/g_numeral.mlg:  2 nts,  4 prods
-----------------------------------------------------
plugins/derive/g_derive.mlg:  1 nts,  1 prods
plugins/extraction/g_extraction.mlg:  20 nts,  29 prods
plugins/funind/g_indfun.mlg:  11 nts,  13 prods
plugins/micromega/g_micromega.mlg:  13 nts,  16 prods
plugins/nsatz/g_nsatz.mlg:  1 nts,  1 prods
plugins/omega/g_omega.mlg:  2 nts,  3 prods
plugins/setoid_ring/g_newring.mlg:  9 nts,  24 prods
plugins/ssr/ssrparser.mlg:  140 nts,  303 prods
plugins/ssr/ssrvernac.mlg:  25 nts,  45 prods
plugins/ssrmatching/g_ssrmatching.mlg:  6 nts,  11 prods
```
(Or examine `wc */*.mlg plugins/*/*.mlg` if you prefer.)

The point of creating a GitHub project is to make the plan visible to others, allowing them to comment?  Or do you think others may be willing to do some of the work on a timely basis?

As a terse summary of the project, I think the main tasks are:
- complete the tool
- revise the mlgs to get a grammar that works well
- add the entire grammar in an appendix
- insert grammar subsets into the documentation, fix dangling grammar links
- revise the doc text as needed to match the grammar

I'll do the tool.  The other parts could be split up by sections (constr, vernac, tactics, ltac, etc.) and done in parallel.

> >The Zimmermann utility

> Huh?

Just a little humor inspired by your suggestion in https://github.com/coq/coq/issues/9310:

> Ideally, given enough annotations, the tool could produce exactly the same grammar as included in the manual and thus verify that the manual is in sync with the parser and be included in CI. Obviously, this should be harder to do, so it could be done as a later step.

**Comment by jfehrle** [comment]
> It would also make sense to move things around so that we strive to have a single grammar (as documented) in a single `mlg` file.

On re-reading this, I suppose you were referring to a combined single grammar for documentation, and not suggesting we put all the .mlgs into one massive file.

**Comment by Zimmi48** [comment]
What I meant was a mlg file would correspond to a given `productionlist`.

**Comment by jfehrle** [comment]
This has been implemented long ago.
